[Pytorch] Add option to CPU Blas GEMM to avoid output downcast

cyrusd98 opened the pull request:
Summary:
Dot product for a single output element consists of 3 steps (both input vectors have elements of type scalar_t):
1. elementwise vector multiply (scalar_t x scalar_t -> opmath_t)
2. vector reduction to a scalar value (opmath_t -> opmath_t)
3. optional downcast if opmath_t != out_t

The current blas kernel performs steps 1 and 2 correctly, but for step 3, it will always downcast to scalar_t even when opmath_t == output_t (and then do an upcast back to output_t), which results in precision loss. This diff fixes the precision loss in the BlasKernel

Test Plan: Attention CI passes

Differential Revision: D75023858

topic: not user facing



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168

Valentine233 reviewed (approved):
LGTM, thanks for the fix.

aditew01 reviewed (approved):
Thanks for the PR. LGTM!

cyrusd98 committed:
[Pytorch] Add option to CPU Blas GEMM to avoid output downcast (#154012)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/154012

Dot product for a single output element consists of 3 steps (both input vectors have elements of type scalar_t):
1. elementwise vector multiply (scalar_t x scalar_t -> opmath_t)
2. vector reduction to a scalar value (opmath_t -> opmath_t)
3. optional downcast if opmath_t != out_t

The current blas kernel performs steps 1 and 2 correctly, but for step 3, it will always downcast to scalar_t even when opmath_t == output_t (and then do an upcast back to output_t), which results in precision loss. This diff fixes the precision loss in the BlasKernel

Test Plan: Attention CI passes

Differential Revision: D75023858

cyrusd98 commented:
@pytorchbot merge

jeanschmidt commented:
forget about this try revert, i followed the wrong link
